iceoryx_posh/source/roudi/port_manager.cpp

```diff
@@ -225,14 +233,175 @@ void PortManager::doDiscoveryForSubscriberPort(SubscriberPortType& subscriberPor
         }
         else
         {
-            // protocol error
+            LogWarn() << "CaPro protocol error for subscriber from runtime '" << subscriberPort.getRuntimeName()
+                      << "' and with service description '" << subscriberPort.getCaProServiceDescription()
+                      << "'! Cannot handle CaProMessageType '" << caproMessage.m_type << "'";
             errorHandler(Error::kPORT_MANAGER__HANDLE_SUBSCRIBER_PORTS_INVALID_CAPRO_MESSAGE,
                          nullptr,
                          iox::ErrorLevel::MODERATE);
         }
     });
 }
 
+void PortManager::destroyClientPort(popo::ClientPortData* const clientPortData) noexcept
+{
+    cxx::Ensures(clientPortData != nullptr && "clientPortData must not be a nullptr");
+
+    // create temporary client ports to orderly shut this client down
+    popo::ClientPortRouDi clientPortRoudi(*clientPortData);
+    popo::ClientPortUser clientPortUser(*clientPortData);
+
+    clientPortRoudi.releaseAllChunks();
```

Comment:
Hmm, it would make more sense to first disconnect and then release all chunks to ensure that there won't arrive new chunks in between. The order is the same for the subscriber. The question is now if there is a reason for this order....

Reply:
Probably just an oversight and I copied the behavior

Reply:
Moving the release of the chunks after the disconnect is not sufficient. It must be moved to the location after the disconnect was announced to the server

Reply:
@elBoberido it was introduced here https://github.com/eclipse-iceoryx/iceoryx/commit/0ae3aa814318e1e0aa060446f14f64259ddf3409 for the subscriber. I thought about it and find no reason why we shouldn't do it the other way round. First ensuring that no one will push to the queue anymore and then clearing the queue seems to be the right order for me. This affects subscriber, client and server

Reply:
I changed it for all ports. I guess it is not a problem because this part of the code should be guarded by a mutex and therefore the order shouldn't matter but once we remove the locks, it should be as you described it.

Reply:
@elBoberido This I do not fully get. A publisher can still deliver data to a connected subscriber after the subscriber made a `releaseAllChunks()` and before we unsubscribe (with a lock). The Roudi code here prevents with a lock that new subscribers/publishers can come in parallel and change the connections

Reply:
@budrus okay, maybe I confused where the lock is. Anyway, the `releaseAllChunks` now happens after dispatching the capro message